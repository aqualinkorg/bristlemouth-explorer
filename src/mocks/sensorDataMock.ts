import { SensorData } from 'src/helpers/types';

export const sensorDataMock: SensorData[] = [
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T20:04:45.000Z',
    sensorPosition: 1,
    units: '°C',
    value: 26.22,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T20:04:45.000Z',
    sensorPosition: 2,
    units: '°C',
    value: 26.1,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T20:24:45.000Z',
    sensorPosition: 1,
    units: '°C',
    value: 26.240000000000002,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T20:24:45.000Z',
    sensorPosition: 2,
    units: '°C',
    value: 26.119999999999997,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T20:44:45.000Z',
    sensorPosition: 1,
    units: '°C',
    value: 26.259999999999998,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T20:44:45.000Z',
    sensorPosition: 2,
    units: '°C',
    value: 26.14,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T21:04:45.000Z',
    sensorPosition: 1,
    units: '°C',
    value: 26.300000000000004,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T21:04:45.000Z',
    sensorPosition: 2,
    units: '°C',
    value: 26.160000000000004,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T21:24:45.000Z',
    sensorPosition: 1,
    units: '°C',
    value: 26.340000000000003,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
  {
    latitude: 20.78995,
    longitude: -156.5069833,
    timestamp: '2023-09-13T21:24:45.000Z',
    sensorPosition: 2,
    units: '°C',
    value: 26.200000000000003,
    unit_type: 'temperature',
    data_type_name: 'sofar_temperature_12bits',
  },
];
