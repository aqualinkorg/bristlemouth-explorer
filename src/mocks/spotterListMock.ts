import { Spotter } from 'src/helpers/types';

export const spotterListMock: Spotter[] = [
  {
    name: 'Totoro',
    spotterId: 'SPOT-300434063450120',
  },
  {
    name: 'USA W01',
    spotterId: 'SPOT-0514',
  },
  {
    name: null,
    spotterId: 'SPOT-1034',
  },
];
