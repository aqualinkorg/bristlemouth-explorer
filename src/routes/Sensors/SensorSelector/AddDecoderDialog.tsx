import {
  Button,
  Dialog,
  DialogActions,
  DialogContent,
  DialogContentText,
  DialogTitle,
  Divider,
  MenuItem,
  Paper,
  Select,
  Stack,
  TextField,
  styled,
} from '@mui/material';
import React from 'react';
import {
  DecoderConfig,
  EncodedDataType,
  SensorStruct,
} from 'src/helpers/types';
import AddIcon from '@mui/icons-material/Add';
import { useAppDispatch } from 'src/store/hooks';
import {
  setSettings,
  settingsSelector,
} from 'src/store/settings/settingsSlice';
import { useSelector } from 'react-redux';

const StyledDivider = styled(Divider)(() => ({
  marginBottom: '2rem',
}));

interface AddDecoderDialogProps {
  open: boolean;
  onClose: () => void;
}

function AddDecoderDialog({ open, onClose }: AddDecoderDialogProps) {
  const dispatch = useAppDispatch();

  const appSettings = useSelector(settingsSelector);

  const [decoderName, setDecoderName] = React.useState<string>('');
  const [decoderConfig, setDecoderConfig] = React.useState<DecoderConfig>([]);

  function handleAddSensor() {
    setDecoderConfig((prev) => [...prev, { name: '', struct: [] }]);
  }

  function handleAddStructKey(sensorName: string) {
    setDecoderConfig((prev) =>
      prev.reduce((acc, curr) => {
        if (curr.name !== sensorName) return [...acc, curr];
        else {
          const newStruct: SensorStruct = [
            ...curr.struct,
            { key: '', dataType: 'uint16_t' },
          ];
          return [
            ...acc,
            {
              name: curr.name,
              struct: newStruct,
            },
          ];
        }
      }, [] as DecoderConfig),
    );
  }

  function handleChangeSensorName(value: string, index: number) {
    setDecoderConfig((prev) => {
      // eslint-disable-next-line fp/no-mutation
      prev[index].name = value;
      return [...prev];
    });
  }

  function handleChangeSensorStructName(value: string, i: number, j: number) {
    setDecoderConfig((prev) => {
      // eslint-disable-next-line fp/no-mutation
      prev[i].struct[j].key = value;
      return [...prev];
    });
  }

  function handleChangeSensorStructDataType(
    value: string,
    i: number,
    j: number,
  ) {
    setDecoderConfig((prev) => {
      // eslint-disable-next-line fp/no-mutation
      prev[i].struct[j].dataType = value as EncodedDataType;
      return [...prev];
    });
  }

  function saveAndClose() {
    dispatch(
      setSettings({
        userDefinedDecoders: [
          ...(appSettings.userDefinedDecoders || []),
          { name: decoderName, config: decoderConfig },
        ],
      }),
    );
    onClose();
  }

  return (
    <Dialog
      open={open}
      onClose={onClose}
      PaperComponent={Paper}
      aria-labelledby="draggable-dialog-title"
    >
      <DialogTitle style={{ cursor: 'move' }} id="draggable-dialog-title">
        Create a decoder
      </DialogTitle>
      <DialogContent>
        <DialogContentText>
          Create a decoder configuration. It will be able saved locally an you
          will be able to re use it the next time you visit this website.
        </DialogContentText>
        <Stack paddingTop="2rem">
          <TextField
            value={decoderName}
            onChange={(e) => setDecoderName(e.target.value)}
            label="decoder name"
            variant="outlined"
          />
          <Divider style={{ paddingTop: '2rem' }} />
          <Stack paddingTop="2rem">
            {decoderConfig.map((config, i) => (
              <Stack gap="2rem">
                <TextField
                  value={config.name}
                  label="sensor name"
                  onChange={(e) => handleChangeSensorName(e.target.value, i)}
                  variant="outlined"
                />
                {config.struct.map((struct, j) => (
                  <Stack gap="1rem">
                    <TextField
                      value={struct.key}
                      onChange={(e) =>
                        handleChangeSensorStructName(e.target.value, i, j)
                      }
                      label={`key${j}`}
                      variant="outlined"
                    />
                    <Select
                      value={struct.dataType}
                      onChange={(e) =>
                        handleChangeSensorStructDataType(e.target.value, i, j)
                      }
                      label="data type"
                    >
                      <MenuItem value="uint16_t">uint16_t</MenuItem>
                      <MenuItem value="float">float</MenuItem>
                      <MenuItem value="double">double</MenuItem>
                    </Select>
                  </Stack>
                ))}
                <Button
                  startIcon={<AddIcon />}
                  variant="outlined"
                  onClick={() => handleAddStructKey(config.name)}
                  style={{ marginTop: '1rem' }}
                >
                  Add struct key
                </Button>
                <StyledDivider />
              </Stack>
            ))}
            <Button
              startIcon={<AddIcon />}
              onClick={() => handleAddSensor()}
              variant="outlined"
              style={{ marginTop: '1rem' }}
            >
              Add sensor
            </Button>
          </Stack>
        </Stack>
      </DialogContent>
      <DialogActions>
        <Button autoFocus onClick={onClose}>
          Cancel
        </Button>
        <Button onClick={saveAndClose}>Create</Button>
      </DialogActions>
    </Dialog>
  );
}

export default AddDecoderDialog;
