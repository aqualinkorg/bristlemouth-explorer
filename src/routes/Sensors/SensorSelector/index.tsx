import React from 'react';
import {
  Button,
  FormControl,
  Link,
  MenuItem,
  Paper,
  Select,
  Stack,
  ToggleButton,
  ToggleButtonGroup,
  Tooltip,
  Typography,
  styled,
} from '@mui/material';
import { AdapterLuxon } from '@mui/x-date-pickers/AdapterLuxon';
import { LocalizationProvider, DatePicker } from '@mui/x-date-pickers';
import { useSelector } from 'react-redux';
import {
  sensorDataRequest,
  sensorDataSelector,
  spottersListSelector,
} from 'src/store/spotters/spottersSlice';
import { Spotter } from 'src/helpers/types';
import { DateTime } from 'luxon';
import { useAppDispatch } from 'src/store/hooks';
import {
  setSettings,
  settingsSelector,
} from 'src/store/settings/settingsSlice';

const PaperContainer = styled(Paper)(({ theme }) => ({
  width: '20rem',
  height: '60vh',
  margin: '0.5rem 0.5rem 0.5rem 1rem',
  padding: '1rem',
  flexShrink: 0,
  borderRadius: theme.spacing(1),
}));

const StyledButton = styled(Button)(() => ({
  textTransform: 'none',
  borderRadius: '50px',
}));

function SensorSelector() {
  const dispatch = useAppDispatch();

  const spottersList = useSelector(spottersListSelector);
  const appSettings = useSelector(settingsSelector);
  const sensorData = useSelector(sensorDataSelector);

  const [timestampFormat, setTimestampFormat] = React.useState<'user' | 'utc'>(
    appSettings.timestampFormat || 'utc',
  );
  const [startDate, setStartDate] = React.useState<DateTime | null>(
    appSettings.spotterDataStartDate
      ? DateTime.fromISO(appSettings.spotterDataStartDate)
      : DateTime.now().minus({ month: 1 }),
  );
  const [endDate, setEndDate] = React.useState<DateTime | null>(
    appSettings.spotterDataEndDate
      ? DateTime.fromISO(appSettings.spotterDataEndDate)
      : DateTime.now(),
  );
  const [selectedSpotter, setSelectedSpotter] = React.useState<Spotter | null>(
    appSettings.selectedSpotter || null,
  );
  const [availableNodeIds, setAvailableNodeIds] = React.useState<string[]>([]);
  const [nodeId, setNodeId] = React.useState<string>('');
  const [decoder, setDecoder] = React.useState<string>(
    appSettings.decoder || '',
  );

  const handleChangeSpotter = (spotterId: string) => {
    const spotter = spottersList.find((x) => x.spotterId === spotterId);
    if (spotter) setSelectedSpotter(spotter);
  };

  async function handleRefresh(updateSettings: boolean) {
    if (selectedSpotter?.spotterId && appSettings.sofarApiToken)
      dispatch(
        sensorDataRequest({
          token: appSettings.sofarApiToken,
          spotterId: selectedSpotter.spotterId,
          startDate: startDate?.startOf('day').toISO() || undefined,
          endDate: endDate?.endOf('day').toISO() || undefined,
        }),
      );

    if (updateSettings)
      dispatch(
        setSettings({
          timestampFormat,
          spotterDataStartDate: startDate?.toISO(),
          spotterDataEndDate: endDate?.toISO(),
          selectedSpotter,
          spotterNodeId: nodeId,
          decoder,
        }),
      );
  }

  // automatically fetch information on first render
  React.useEffect(() => {
    handleRefresh(false);
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, []);

  React.useEffect(() => {
    if (selectedSpotter === null && spottersList.length !== 0) {
      setSelectedSpotter(spottersList[0]);
    }
  }, [setSelectedSpotter, selectedSpotter, spottersList]);

  React.useEffect(() => {
    if (sensorData.length === 0) return;

    const uniqueNodeIds = [
      ...new Map(sensorData.map((x) => [x.bristlemouth_node_id, null])).keys(),
    ];

    setAvailableNodeIds(uniqueNodeIds);
  }, [sensorData]);

  /**
   * Initialize the 'nodeId' after obtaining 'availableNodeIds' to prevent
   * MUI warnings related to out-of-range values for the 'Select' component.
   */
  React.useEffect(() => {
    if (!availableNodeIds) return;
    const node = availableNodeIds.find((x) => x === appSettings.spotterNodeId);
    if (node !== undefined) setNodeId(node);

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [availableNodeIds]);

  return (
    <LocalizationProvider dateAdapter={AdapterLuxon}>
      <PaperContainer>
        <Stack justifyContent="space-between" height="100%" overflow="scroll">
          <Stack gap="1.5rem">
            <Stack gap="0.5rem">
              <Typography fontWeight="bold">SPOT ID</Typography>
              <FormControl size="small">
                <Select
                  value={
                    spottersList.length > 0 ? selectedSpotter?.spotterId : ''
                  }
                  onChange={(e) => handleChangeSpotter(e.target.value)}
                >
                  {spottersList.map((x) => (
                    <MenuItem value={x.spotterId} key={x.spotterId}>
                      {x.spotterId}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </Stack>

            <Stack gap="0.5rem">
              <Typography fontWeight="bold">Date range</Typography>
              <Stack direction="row" justifyContent="space-between" gap="1rem">
                <Stack>
                  <Typography>Start date</Typography>
                  <DatePicker
                    maxDate={endDate}
                    slotProps={{ textField: { size: 'small' } }}
                    disabled={spottersList.length === 0}
                    timezone="UTC"
                    value={startDate}
                    onChange={(newValue) => setStartDate(newValue)}
                  />
                </Stack>
                <Stack>
                  <Typography>End date</Typography>
                  <DatePicker
                    minDate={startDate}
                    slotProps={{ textField: { size: 'small' } }}
                    disabled={spottersList.length === 0}
                    timezone="UTC"
                    value={endDate}
                    onChange={(newValue) => setEndDate(newValue)}
                  />
                </Stack>
              </Stack>
            </Stack>

            <Stack gap="0.5rem">
              <Typography fontWeight="bold">Node ID</Typography>
              <FormControl size="small">
                <Select
                  value={nodeId}
                  onChange={(e) => {
                    const { value } = e.target;
                    setNodeId(value);
                    dispatch(
                      setSettings({
                        spotterNodeId: value,
                      }),
                    );
                  }}
                >
                  <MenuItem value="">&nbsp;</MenuItem>
                  {availableNodeIds.map((x) => (
                    <MenuItem key={x} value={x}>
                      {x}
                    </MenuItem>
                  ))}
                </Select>
              </FormControl>
            </Stack>

            <Stack gap="0.5rem">
              <Typography fontWeight="bold">Decoder</Typography>
              <Tooltip title="coming soon">
                <FormControl size="small">
                  <Select
                    disabled
                    value={decoder}
                    onChange={(e) => {
                      const { value } = e.target;
                      setDecoder(value);
                      dispatch(
                        setSettings({
                          decoder: value,
                        }),
                      );
                    }}
                  >
                    <MenuItem value="">&nbsp;</MenuItem>
                  </Select>
                </FormControl>
              </Tooltip>
            </Stack>

            <Stack gap="0.5rem">
              <Typography fontWeight="bold">Timestamp</Typography>
              <ToggleButtonGroup
                color="primary"
                exclusive
                value={timestampFormat}
                onChange={(_, val) => {
                  if (val) {
                    setTimestampFormat(val);
                    dispatch(
                      setSettings({
                        timestampFormat: val,
                      }),
                    );
                  }
                }}
              >
                <ToggleButton value="user">User</ToggleButton>
                <ToggleButton value="utc">UTC</ToggleButton>
              </ToggleButtonGroup>
            </Stack>

            <StyledButton
              disabled={
                !(selectedSpotter?.spotterId && appSettings.sofarApiToken)
              }
              variant="contained"
              onClick={() => handleRefresh(true)}
            >
              <Typography
                lineHeight="2rem"
                fontWeight="bold"
                color="white"
                height="2rem"
              >
                Refresh
              </Typography>
            </StyledButton>
          </Stack>
          <Link paddingTop="0.5rem" href="/">
            Change your API token
          </Link>
        </Stack>
      </PaperContainer>
    </LocalizationProvider>
  );
}

export default SensorSelector;
