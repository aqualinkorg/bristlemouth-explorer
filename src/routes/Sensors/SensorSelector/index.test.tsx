import { render } from '@testing-library/react';
import SensorSelector from '.';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { BrowserRouter } from 'react-router-dom';
import { spotterListMock } from 'src/mocks/spotterListMock';
import { sensorDataMock } from 'src/mocks/sensorDataMock';

const mockStore = configureStore([]);

const store = mockStore({
  spotters: {
    list: spotterListMock,
    sensorData: sensorDataMock,
    sensorDataLoading: false,
    spottersRequestLoading: false,
    error: null,
  },
  settings: {
    selectedSpotter: spotterListMock[0],
    sofarApiToken: undefined,
    spotterDataStartDate: '2023-09-20T03:00:00.000+03:00',
    spotterDataEndDate: '2023-09-20T03:00:00.000+03:00',
    spotterNodeId: undefined,
    decoder: undefined,
    timestamp: undefined,
  },
});

store.dispatch = jest.fn();

it('renders as expected', () => {
  const { container } = render(
    <Provider store={store}>
      <BrowserRouter>
        <SensorSelector />
      </BrowserRouter>
    </Provider>,
  );
  expect(container).toMatchSnapshot();
});
