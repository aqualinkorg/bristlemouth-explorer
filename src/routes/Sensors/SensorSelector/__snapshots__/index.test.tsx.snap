// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders as expected 1`] = `
<div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 css-uhoyx7-MuiPaper-root"
  >
    <div
      class="MuiStack-root css-16slztd-MuiStack-root"
    >
      <div
        class="MuiStack-root css-qre78q-MuiStack-root"
      >
        <div
          class="MuiStack-root css-5u2gm5-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            SPOT ID
          </p>
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                role="button"
                tabindex="0"
              >
                SPOT-300434063450120
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="SPOT-300434063450120"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-5u2gm5-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Date range
          </p>
          <div
            class="MuiStack-root css-xp3pc1-MuiStack-root"
          >
            <div
              class="MuiStack-root css-nen11g-MuiStack-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                Start date
              </p>
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    autocomplete="off"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r1:"
                    inputmode="text"
                    placeholder="MM/DD/YYYY"
                    type="text"
                    value="09/20/2023"
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
                  >
                    <button
                      aria-label="Choose date, selected date is Sep 19, 2023"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CalendarIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
            <div
              class="MuiStack-root css-nen11g-MuiStack-root"
            >
              <p
                class="MuiTypography-root MuiTypography-body1 css-ahj2mt-MuiTypography-root"
              >
                End date
              </p>
              <div
                class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
              >
                <div
                  class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
                >
                  <input
                    aria-invalid="false"
                    autocomplete="off"
                    class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall MuiInputBase-inputAdornedEnd css-19qh8xo-MuiInputBase-input-MuiOutlinedInput-input"
                    id=":r3:"
                    inputmode="text"
                    placeholder="MM/DD/YYYY"
                    type="text"
                    value="09/20/2023"
                  />
                  <div
                    class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeSmall css-1laqsz7-MuiInputAdornment-root"
                  >
                    <button
                      aria-label="Choose date, selected date is Sep 19, 2023"
                      class="MuiButtonBase-root MuiIconButton-root MuiIconButton-edgeEnd MuiIconButton-sizeMedium css-1yq5fb3-MuiButtonBase-root-MuiIconButton-root"
                      tabindex="0"
                      type="button"
                    >
                      <svg
                        aria-hidden="true"
                        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                        data-testid="CalendarIcon"
                        focusable="false"
                        viewBox="0 0 24 24"
                      >
                        <path
                          d="M17 12h-5v5h5v-5zM16 1v2H8V1H6v2H5c-1.11 0-1.99.9-1.99 2L3 19c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2h-1V1h-2zm3 18H5V8h14v11z"
                        />
                      </svg>
                      <span
                        class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
                      />
                    </button>
                  </div>
                  <fieldset
                    aria-hidden="true"
                    class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                  >
                    <legend
                      class="css-ihdtdm"
                    >
                      <span
                        class="notranslate"
                      >
                        ​
                      </span>
                    </legend>
                  </fieldset>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-5u2gm5-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Node ID
          </p>
          <div
            class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
          >
            <div
              class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
            >
              <div
                aria-expanded="false"
                aria-haspopup="listbox"
                class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                role="button"
                tabindex="0"
              >
                0xadef442a8e274b78
              </div>
              <input
                aria-hidden="true"
                aria-invalid="false"
                class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                tabindex="-1"
                value="0xadef442a8e274b78"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                data-testid="ArrowDropDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
              >
                <legend
                  class="css-ihdtdm"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </legend>
              </fieldset>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-5u2gm5-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Decoder
          </p>
          <div
            class="MuiStack-root css-11n39z-MuiStack-root"
          >
            <div
              class="MuiFormControl-root MuiFormControl-fullWidth css-q8hpuo-MuiFormControl-root"
            >
              <div
                class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-sizeSmall  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
              >
                <div
                  aria-expanded="false"
                  aria-haspopup="listbox"
                  class="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputSizeSmall css-jedpe8-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
                  role="button"
                  tabindex="0"
                >
                  <span
                    class="notranslate"
                  >
                    ​
                  </span>
                </div>
                <input
                  aria-hidden="true"
                  aria-invalid="false"
                  class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
                  tabindex="-1"
                  value=""
                />
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
                  data-testid="ArrowDropDownIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M7 10l5 5 5-5z"
                  />
                </svg>
                <fieldset
                  aria-hidden="true"
                  class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                >
                  <legend
                    class="css-ihdtdm"
                  >
                    <span
                      class="notranslate"
                    >
                      ​
                    </span>
                  </legend>
                </fieldset>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiStack-root css-5u2gm5-MuiStack-root"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-ndaqpv-MuiTypography-root"
          >
            Timestamp
          </p>
          <div
            class="MuiToggleButtonGroup-root css-1gjgmky-MuiToggleButtonGroup-root"
            role="group"
          >
            <button
              aria-pressed="false"
              class="MuiButtonBase-root MuiToggleButton-root MuiToggleButton-sizeMedium MuiToggleButton-primary MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-d9c359-MuiButtonBase-root-MuiToggleButton-root"
              tabindex="0"
              type="button"
              value="user"
            >
              User
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
            <button
              aria-pressed="true"
              class="MuiButtonBase-root MuiToggleButton-root Mui-selected MuiToggleButton-sizeMedium MuiToggleButton-primary MuiToggleButtonGroup-grouped MuiToggleButtonGroup-groupedHorizontal css-d9c359-MuiButtonBase-root-MuiToggleButton-root"
              tabindex="0"
              type="button"
              value="utc"
            >
              UTC
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </div>
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-disabled MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-sizeMedium MuiButton-containedSizeMedium css-1ow0rcz-MuiButtonBase-root-MuiButton-root"
          disabled=""
          tabindex="-1"
          type="button"
        >
          <p
            class="MuiTypography-root MuiTypography-body1 css-jlrj90-MuiTypography-root"
          >
            Refresh
          </p>
        </button>
      </div>
      <a
        class="MuiTypography-root MuiTypography-inherit MuiLink-root MuiLink-underlineAlways css-2gjjad-MuiTypography-root-MuiLink-root"
        href="/"
      >
        Change your API token
      </a>
    </div>
  </div>
</div>
`;
